### PR TITLE
Use the V7 switch binder if the switch contains any new features (VS crash fix)

### PR DESCRIPTION
**Pull request by gafter**
**Customer scenario**

In a project set to use language version 6, start typing a switch statement inside an existing method, where auto-brace-insertion is off and there is a var-declared variable following the switch being typed. The problem occurs at this stage of typing:
``` c#
switch (e)
{
    case // cursor here

var x = y;
M(x);
```
After typing `case`, VS crashes.

**Bugs this fixes:**

Fixes https://devdiv.visualstudio.com/DefaultCollection/DevDiv/_workitems?_a=edit&id=388743
and numerous Watson crashes linked to that.

**Workarounds, if any**

Turn on automatic brace insertion.

**Risk**

Low. The fix is to bind code that appears to use C# 7 constructs as if they were written for C# 7 (after giving an error about feature not being available in C# 6). This is a single-line change.

**Performance impact**

No performance impact expected. New code paths are only exercised for erroneous code.

**Is this a regression from a previous update?**

Yes, VS did not crash in similar scenarios in VS2015 and earlier. Problem was introduced in VS2017 with the introduction of pattern-matching.

**Root cause analysis:**

Compiler code author assumed that syntax errors would preclude binding. But that is not the case; we bind even erroneous code for IDE scenarios.

**How was the bug found?**

Watson failures in customer use.

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Binder/SwitchBinder.cs [hunk 1]
```diff
@@ -36,11 +36,25 @@ internal static SwitchBinder Create(Binder next, SwitchStatementSyntax switchSyn
                 // lead to a translation that fully supports edit-and-continue, so it delegates to the C# 6
                 // binder when it can. The "testV7SwitchBinder" feature flag forces the use of the C# 7 switch binder
                 // for all operations; we use it to enhance test coverage.
-                (parseOptions?.IsFeatureEnabled(MessageID.IDS_FeaturePatternMatching) != false || parseOptions?.Features.ContainsKey("testV7SwitchBinder") != false)
+                (parseOptions?.IsFeatureEnabled(MessageID.IDS_FeaturePatternMatching) != false ||
+                 parseOptions?.Features.ContainsKey("testV7SwitchBinder") != false ||
+                 switchSyntax.HasErrors && HasPatternSwitchSyntax(switchSyntax))
```
Comment (gafter) [inline src/Compilers/CSharp/Portable/Binder/SwitchBinder.cs]: This is the key to the fix: use a V7 switch binder if there are any patterns in the switch, even in language version 6.

**Comment by gafter** [comment]
/CC @MeiChin-Tsai for ask mode approval
/CC @dotnet/roslyn-compiler May I please have a couple of reviews of this tiny crash fix?

File: src/Compilers/CSharp/Test/Semantic/Semantics/PatternSwitchTests.cs [hunk 2]
```diff
@@ -3102,5 +3099,84 @@ static void Main(string[] args)
...
+                //             case var q:
+                Diagnostic(ErrorCode.ERR_FeatureNotAvailableInVersion6, "case var q:").WithArguments("pattern matching", "7").WithLocation(7, 13)
+                );
+            var tree = compilation.SyntaxTrees[0];
+            var model = compilation.GetSemanticModel(tree);
+            var node = tree.GetRoot().DescendantNodes()
+                .OfType<IdentifierNameSyntax>()
+                .Where(n => n.Identifier.ValueText == "q" && n.ToFullString().Contains("/*BIND*/"))
+                .Single();
+            var type = model.GetTypeInfo(node);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/PatternSwitchTests.cs]: >type [](start = 16, length = 4)

Consider asserting content of this structure.

**Comment by MeiChin-Tsai** [comment]
Approved pending all gates are green and comments are addressed. thx.

**Comment by gafter** [comment]
retest vs-integration_prtest please

**Comment by gafter** [comment]
retest windows_debug_unit32_prtest please
retest windows_debug_unit64_prtest please

**Comment by gafter** [comment]
retest windows_debug_unit32_prtest please

**Comment by gafter** [comment]
retest windows_debug_unit64_prtest please

**Comment by gafter** [comment]
retest microbuild_prtest please

**Comment by gafter** [comment]
/CC @dotnet/roslyn-compiler May I please have a second review of this tiny crash fix?

**Comment by cston** [comment]
LGTM

**Comment by gafter** [comment]
retest windows_debug_vs-integration_prtest please

**Comment by gafter** [comment]
retest windows_release_unit32_prtest please

**Pull request merged**